building_sim_plugins/building_plugins_common/include/building_sim_common/crowd_simulator_common.hpp

```diff
@@ -0,0 +1,208 @@
+#ifndef BUILDING_SIM_COMMON__CROWD_SIMULATOR_COMMON_HPP
+#define BUILDING_SIM_COMMON__CROWD_SIMULATOR_COMMON_HPP
+
+#include <functional>
+#include <list>
+#include <queue>
+#include <memory>
+
+#include <MengeCore/Runtime/SimulatorDB.h>
+#include <MengeCore/Orca/ORCADBEntry.h>
+#include <MengeCore/Orca/ORCASimulator.h>
+#include <MengeCore/PluginEngine/CorePluginEngine.h>
+
+namespace crowd_simulator {
+
+using AgentPtr = std::shared_ptr<Menge::Agents::BaseAgent>;
+//===============================================================
+/*
+* class AgentPose3d
+* replace the common interface of ignition::math::Pose3d, which might apply with different version in different plugins
+*/
+class  AgentPose3d{
+
+public:
+  AgentPose3d();
+  AgentPose3d(double& x, double& y, double& z, double& pitch, double& roll, double& yaw);
+
+  ~AgentPose3d();
+
+  double X() const;
+  double Y() const;
+  double Z() const;
+  double Pitch() const;
+  double Roll() const;
+  double Yaw() const;
+
+  double& X();
+  double& Y();
+  double& Z();
+  double& Pitch();
+  double& Roll();
+  double& Yaw();
+
+  void X(const double& x);
+  void Y(const double& y);
+  void Z(const double& z);
+  void Pitch(const double& pitch);
+  void Roll(const double& roll);
+  void Yaw(const double& yaw);
+
+private:
+  double _x;
+  double _y;
+  double _z;
+  double _pitch;
+  double _roll;
+  double _yaw;
+
+};
+
+//================================================================
+/*
+* class MengeHandle
+*/
+class MengeHandle
+{
+public:
+
+  MengeHandle(const std::string& resourcePath,
+    const std::string& behaviorFile,
+    const std::string& sceneFile,
+    const float simTimeStep = 0.0
+  )
+  : _resourcePath(resourcePath),
+    _behaviorFile(behaviorFile),
+    _sceneFile(sceneFile),
+    _simTimeStep(simTimeStep),
+    _agentCount(0)
+  {
+
+    _behaviorFile = this->_ResourceFilePath(_behaviorFile);
+    _sceneFile = this->_ResourceFilePath(_sceneFile);
+
+    if (this->_LoadSimulation())
+    {
+      this->initialized = true;
+    }
+    assert(this->initialized);
```

Comment:
I might be missing something, but it looks like the `initialized` boolean doesn't serve any specific purpose here. 

For ensuring everything is loaded or constructed correctly before proceeding with the simulation plugin, you can consider using factory methods, https://refactoring.guru/design-patterns/factory-method/cpp/example. It is currently used extensively in `rmf_core` as static `make` functions, for example,
https://github.com/osrf/rmf_core/blob/master/rmf_fleet_adapter/include/rmf_fleet_adapter/agv/Adapter.hpp#L36-L80
https://github.com/osrf/rmf_core/blob/master/rmf_fleet_adapter/src/rmf_fleet_adapter/agv/Adapter.cpp#L215-L242

This will allow you to propagate a failure all the way to the plugin and fail the plugin gracefully, instead of needing asserts or boolean `initialized` checks.

Reply:
emm, `MengeHandle` class is a wrap up for the actual menge simulator on the purpose of only exposing the agent position related function. the _sim variable pointing to the actual menge simulator. After that, the CrowdSimInterface will use MengeHandle to access the agent position. `initialized` in Menge will be checked in CrowdSimInterface.

Reply:
Could you point me to the code that checks for it in `CrowdSimInterface`? I can't seem to find it.

Reply:
sorry, this `initialized` variable is not checked in `CrowdSimInterface`. Yeah, I should use the factory pattern. will modify that.